### PR TITLE
Dead letter queue implementation

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Added support for dead letter queue message routing

### CODE REVIEW

File: pulsar/dlq_router.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+			}
+			r.log.Debug("Closed DLQ router")
+			return
+		}
+	}
+}
+
+func (r *dlqRouter) close() {
+	if r.closeCh != nil {
+		r.closeCh <- nil
```
Comment (cckellogg) [inline pulsar/dlq_router.go]: I think this could cause a deadlock for anything but a single topic because there will be multiple calls to close() and only the first call will be able to send on the closeCh because the go routine reading from the closeCh will exit.

We could do something like this:
```
select {
 case r.closeCh <- nil: // this case will only occur when we can send on the channel
default:
// the closeCh is either nil or it's blocked and we can't send on it
}
```

File: pulsar/consumer_impl.go [hunk 2]
```diff
@@ -74,6 +75,11 @@ func newConsumer(client *client, options ConsumerOptions) (Consumer, error) {
 		messageCh = make(chan ConsumerMessage, 10)
 	}
 
+	dlq, err := newDlqRouter(client, options.DLQ)
```
Comment (cckellogg) [inline pulsar/consumer_impl.go]: what are the defaults here? will every topic by default have dlq enabled?

File: pulsar/consumer_partition.go [hunk 3]
```diff
@@ -402,23 +404,27 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
...
-				msgID:       msgID,
-				payLoad:     payload,
+				publishTime:         timeFromUnixTimestampMillis(msgMeta.GetPublishTime()),
+				eventTime:           timeFromUnixTimestampMillis(smm.GetEventTime()),
+				key:                 smm.GetPartitionKey(),
+				properties:          internal.ConvertToStringMap(smm.GetProperties()),
+				topic:               pc.topic,
+				msgID:               msgID,
+				payLoad:             payload,
+				replicationClusters: msgMeta.GetReplicateTo(),
```
Comment (jerrypeng) [inline pulsar/consumer_partition.go]: why is this added?

File: pulsar/dlq_router.go (hunk 1 above)
Comment (merlimat) [reply pulsar/dlq_router.go]: Added select and test case with multi topics

File: pulsar/consumer_impl.go (hunk 2 above)
Comment (merlimat) [reply pulsar/consumer_impl.go]: By default the policy is `nil` and the it will result in no-op

File: pulsar/consumer_partition.go (hunk 3 above)
Comment (merlimat) [reply pulsar/consumer_partition.go]: We need to copy all properties of the original message and noticed that this was not being exposed in the internal struct.
